plugin/pulpcore/plugin/changeset/model.py

```diff
@@ -198,10 +298,51 @@ def save(self):
         Update the DB model to store the downloaded file.
         Then, save in the DB.
         """
-        if self.path:
-            self.model.file = File(open(self.path, mode='rb'))
-            self.model.downloaded = True
-        super(RemoteArtifact, self).save()
+        artifact = None
+
+        if not self.fetched:
+            if self._path:  # downloaded
+                try:
+                    with transaction.atomic():
+                        self.model = Artifact(
+                            file=File(open(self._path, mode='rb')),
+                            **self.monitor.dict())
+                        self.model.save()
+                except IntegrityError:
+                    q = self.artifact_q()
+                    self.model = Artifact.objects.get(q)
+                finally:
+                    artifact = self.model
+        else:
+            artifact = self.model
+
+        try:
+            with transaction.atomic():
+                content_artifact = ContentArtifact(
+                    relative_path=self.relative_path,
+                    content=self.content.model,
+                    artifact=artifact)
+                content_artifact.save()
+        except IntegrityError:
+            content_artifact = ContentArtifact.objects.get(
+                relative_path=self.relative_path,
+                content=self.content.model)
+            if self.fetched:
+                content_artifact.artifact = artifact
+                content_artifact.save()
+
+        try:
+            with transaction.atomic():
+                deferred_artifact = DeferredArtifact(
+                    url=self.url,
+                    importer=self.importer,
+                    content_artifact=content_artifact,
+                    size=self.model.size,
+                    **{f: getattr(self.model, f) for f in Artifact.DIGEST_FIELDS})
+                deferred_artifact.save()
+        except IntegrityError:
+            # TODO: update?
```

Comment:
Looking at the ``DeferredArtifact`` model, it doesn't have any uniqueness constraints beyond the default ``pk`` one. So given that, I don't think you'd ever get an IntegrityError, nor would you need to do this in a transaction.

That doesn't seem intuitive or correct though. We need to think about uniqueness first on the model, and then come back here to think about collisions.

I suspect it should be unique based on the importer and the content_artifact, and that if any other attributes change, the DB record should be updated.

Thoughts @jortel and @dkliban ?